### ISSUE TITLE
Status code to use in a response indicating rejection of a cross-origin access

### ISSUE THREAD

**Issue by tyoshino** [whatwg/fetch#172]
**tl;dr:** see the title.

**Long and trivial version:**

The CORS protocol can work independently from the response representation. I.e. attaching/omitting the `Access-Control-Allow-Origin` header can instruct UAs whether or not to allow the script to see the response without making any change on the other headers and status-line.

So, to define the CORS algorith, technically, we don't have to define any dedicated HTTP status code for representing rejection due to a bad Origin value. Actually, the current spec doesn't say anything (neither requirement nor recommendation) about a status code to include in a response representing cross-origin access rejection.

Does this design imply that the server should:
- serve the contents anyway
- but decide whether or not to include `Allow-Control-Allow-Origin` to instruct the UA how to handle the response in CORS enabled case
- and then delegate the final decision to the UA. It's possible that the UA doesn't use the `Allow-Control-Allow-Origin` in case the request was issued in the `no-CORS` mode.
  - this (the server sees Origin header but the UA is in `no-CORS` mode) happens since XHR and Fetch works with omit-Origin-header-flag unset.

Regardless the answer is yes or no, in practice, developers want to reject unexpected cross origin requests at minimum cost. We basically want to keep the connection alive, so disconnect is not the best idea. Then, we generate a response w/o `Allow-Control-Allow-Origin`, body. I.e. the response no longer represent any valid response from the application layer. What status code should we put in the response? Maybe 403 is appropriate. But I wonder if there any discussion about this topic has happened in past.

**Comment by mnot** [comment]
In CORS, the enforcement point is the client. If you're looking for how servers should express insufficient credentials when *they* are the enforcement point, 403 would be the logical choice (unless you want to prompt for a different credential, in which case 401 would do the trick).

**Comment by tyoshino** [comment]
Thanks, Mark. So, it seems there's no existing status code that fits this.
https://tools.ietf.org/html/rfc7231#section-6

Maybe the choice developers have is to use some status code ignoring the semantics defined in the spec and but considering safety (avoid caching, etc.).

**Comment by tyoshino** [comment]
Discussed this with @yutakahirano.

When the ServiceWorker is used, we may serve (by calling `respondWith()`) a no-CORS request for loading a sub-resource with a response fetched using a modified request (e.g. changing URL). The modified request's `omit-Origin-header-flag` is unset. Given this, it seems that a server basically shouldn't alter a response to send based on presence of a Origin header. Applying a stricter rule for requests with the Origin header may unexpectedly apply a stricter rule to SW-controlled pages even though either the SW or the controlled page's script cannot see the contents of the no-CORS response.

Now, you see that the via-SW request is not distinguishable from a request issued by CORS enabled fetching. Both have the Origin header. The mode (no-CORS or CORS) is not visible to the server. So, there's no way to serve them differently. If you're going to accept no-CORS requests based on the above discussion, you also have to accept CORS requests.

This suggests that we shouldn't do anything special at the server basically. As Mark said, the enforcement point is the client and only the client knows the mode (CORS/no-CORS). It seems the best thing to do is just serving any request regardless of the Origin presence but setting `Access-Control-Allow-Origin` appropriately based on the Origin value if any.

----

One can give up serving requests w/o the Origin header correctly, and return simplified body-less 4XX response for bad Origin requests, if they really want. Maybe it helps DDoS protection to some extent.

**Comment by mnot** [comment]
Oh, wait -- are you talking about a status code to be generated *by* fetch if there's a cross-origin access failure? If so, 403 is definitely the right one; it's not restricted to use on the origin server. You can add headers (or body) to refine the semantics if you like, but a new status code isn't necessary.

**Comment by annevk** [comment]
No @tyoshino is asking what a server should do if it implements CORS and does not want to give a normal response to a given request. However, as he pointed out the server cannot distinguish a CORS request from other requests since `Origin` can be used for either.

It seems in such scenarios the server can use 400 or 403, but I'm not sure the specification needs to say anything since the client will not act on it necessarily.

(As far as any APIs are concerned whenever the CORS protocol is not honered it is a network error which has a status of 0.)

**Comment by tyoshino** [comment]
Right. About a status code sent by a server.

I'm also not sure if there should be some text in the spec. But it's good to have some normative guideline for server developers. Even an issue, e.g. in this one, MDN, something like FAQ you had on w3.org, etc..., explaining howto would be useful.

**Comment by mnot** [comment]
OK. Generally, we only try to mint new status codes when there's generic HTTP software that's making a decision based upon the status code in isolation.

**Comment by timbl** [comment]
@mnot in " a status code to be generated by fetch if there's a cross-origin access failure? If so, 403 is definitely the right one." Are you suggesting that the fetch function locally should fake an HTTP status code?  That sounds to be like a bad idea -- the caller really needs to understand what went on.  Better surely to make a new series of client-side error code 600 series maybe and allocate them for different kinds of CORS failure.   Of course currently often the status code is 0 for cors problems, and the calling code has to guess what happened -- sometimes ending up wasting time implementing different alternative fall backs in turn.

**Comment by mnot** [comment]
@timbl Yes; 403 can be generated by any server, not just the origin, and effectively Fetch is acting as a server in this case because it's generating a response (or it would be if that were what was being discussed, but I misunderstood it).

Debugging information / context is best to put into headers or the body (although, as always, it's important to be careful about how much of such information is exposed to a possible attacker).

**Comment by annevk** [comment]
I tried to clarify the requirements on a CORS server. Could you maybe review PR #511 @tyoshino?
